bindings/src/main/java/io/pravega/storage/filesystem/FileSystemStorage.java

```diff
@@ -458,5 +466,225 @@ private Void doDelete(SegmentHandle handle) throws IOException {
+                    throw ex;
+                }
+            }
+        }
+
+        private Void replaceExistingFile(String segmentName, BufferView contents) throws IOException, StreamSegmentException {
+            boolean baseExists = super.doExists(segmentName);
+            boolean shouldReseal = baseExists && super.getStreamSegmentInfo(segmentName).isSealed();
+
+            // Check temp file already exists.
+            String tmpSegmentName = getTempSegmentName(segmentName);
+            if (super.doExists(tmpSegmentName)) {
+                if (baseExists) {
+                    // We have both a temp file and a segment file. This is most likely the result of an incomplete replace,
+                    // however we still have the original segment file around. It is safe to delete the temp file now.
+                    log.debug("Incomplete replace operation detected for '{}'. Deleting temp file before new replace attempt.", segmentName);
+                    super.doDelete(super.doOpenWrite(tmpSegmentName));
+                } else {
+                    // Temp file exists, but the segment file does not. This may be the result of an incomplete replace,
+                    // in which case we need to finalize that one to prevent deleting (what could be) the only persisted
+                    // copy of our data.
+                    log.debug("Incomplete replace operation detected for '{}'. Finalizing before new replace attempt.", segmentName);
+                    finalizeRename(tmpSegmentName, segmentName);
+                }
+            } else if (!baseExists) {
+                throw new StreamSegmentNotExistsException(segmentName);
+            }
+
+            // Write given contents to temp file.
+            val tmpHandle = super.doCreate(tmpSegmentName);
+            try {
+                super.doWrite(tmpHandle, 0, contents.getReader(), contents.getLength());
+                if (shouldReseal) {
+                    super.doSeal(tmpHandle);
+                }
+            } catch (Exception ex) {
+                log.warn("Unable to write to temporary file when attempting to replace '{}'. Original file has not been touched. Cleaning up.",
+                        segmentName, ex);
+                super.doDelete(tmpHandle);
+                throw ex;
+            }
+
+            // Sanity check #1 (only executes in tests).
+            assert super.doGetStreamSegmentInfo(tmpSegmentName).getLength() == contents.getLength();
+
+            // Rename file. After this is done, the replace operation is complete.
+            finalizeRename(tmpSegmentName, segmentName);
+
+            // Sanity check #2 (only executes in tests).
+            assert super.doGetStreamSegmentInfo(segmentName).getLength() == contents.getLength();
+            return null;
+        }
+
+        @SneakyThrows(StreamSegmentReplaceException.class)
+        private void finalizeRename(String fromSegmentName, String toSegmentName) {
```

Comment:
Maybe add a debug or trace log to record the rename.

Reply:
Done.